### PR TITLE
Methods for adding a `CircBox` registerwise

### LINKED ISSUE #1057: Allow `CircBox` to contain arbitrary `UnitID`
Add an optional `Dict[UnitID, UnitID]` argument to `Circuit.add_circbox()` to specify the wiring. (By default wiring is done in lexicographic order.)

### LINKED ISSUE #1059: Improve error message for `CircBox` creation from circuit with non-default register
"Function only allowed for simple circuits (single register)" isn't clear.

### PR DESCRIPTION
**Pull request by cqc-alec**
# Description

Drop the restriction that a `CircBox` must have default registers. Add a couple of methods for adding a `CircBox` by matching up registers instead of individual qubits.

# Related issues

Closes #1057 .
Closes #1059 .

# Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented hard-to-understand parts of my code.
- [x] I have made corresponding changes to the public API documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have updated the changelog with any user-facing changes.

### CODE REVIEW

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 1]
```diff
@@ -517,11 +524,144 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
+            }
+            for (const auto &pair : box_creg_sizes) {
+              if (circ_creg_sizes.at(cregmap.at(pair.first)) != pair.second) {
+                throw CircuitInvalidity("Size mismatch in bit registers");
+              }
+            }
+
+            // Populate the arguments (qubits then bits):
+            std::vector<UnitID> args;
+            for (const Qubit &qb : box_qubits) {
```
Comment (yao-cqc) [inline pytket/binders/circuit/Circuit/add_op.cpp:638]: It seems like for this to work, we need both `box_qubits = box_circ->all_qubits()` and `replacement.flatten_registers()` to handle `UnitID`s in lexicographic order. I'm wondering if we could explain this better, maybe with a comment.

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 2]
```diff
@@ -517,11 +524,144 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
+            }
+            return add_box_method(
+                circ, std::make_shared<CircBox>(box), args, kwargs);
+          },
+          "Append a :py:class:`CircBox` to the circuit, wiring whole registers "
+          "together."
+          "\n\n:param circbox: The box to append"
+          "\n:param qregs: Sequence of :py:class:`QubitRegister` from the "
+          "outer :py:class:`Circuit`, in order of corresponding registers in "
+          "the :py:class:`CircBox`"
```
Comment (yao-cqc) [inline pytket/binders/circuit/Circuit/add_op.cpp:561]: Should we specify _lexicographic order_ for clarity?

**Review by yao-cqc — COMMENTED** [review]
Looks good! Just two minor comments.

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 1 above)
Comment (cqc-alec) [reply pytket/binders/circuit/Circuit/add_op.cpp:643]: Done.

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 2 above)
Comment (cqc-alec) [reply pytket/binders/circuit/Circuit/add_op.cpp]: Done.

**Review by yao-cqc — APPROVED** [review]

**Pull request merged**